cost/downsize_instance/downsize-instance.pt

```diff
@@ -74,70 +89,76 @@ datasource "ds_instances" do
   field "instance_type", val(iter_item, "instance_type")
 end
 
-datasource "ds_instances_tags" do
+datasource "ds_instances_16" do
+  iterate(@clouds)
   request do
     auth $auth_rs
-    verb "POST"
+    verb "GET"
     host rs_cm_host
-    path "/api/tags/by_resource"
-    header "X-Api-Version", "1.5"
-    body_field "resource_hrefs", hrefs(@instances)
+    path join([href(iter_item), "/instances"])
+    header "X-Api-Version", "1.6"
+    header "X-Account", to_s(rs_project_id)
+  end
+  result do
+    encoding "json"
+    collect jmes_path(response, "[*]") do
+      field "href", jmes_path(col_item, "href")
+      field "name", jmes_path(col_item,"name")
+      field "description", jmes_path(col_item, "description")
+      field "legacy_id", to_n(jmes_path(col_item,"legacy_id"))
+      field "instance_type", jmes_path(col_item, "links.instance_type.name")
+      field "region", jmes_path(col_item, "links.cloud.region")
+      field "cloud_type", jmes_path(col_item, "links.cloud.cloud_type")
+    end
   end
 end
 
+datasource "ds_instance_type" do
+  request do
+    verb "GET"
+    host "raw.githubusercontent.com"
+    path "/rightscale/policy_templates/PSASSET-863/cost/downsize_instance/instance_types.json"
+    header "User-Agent", "RS Policies"
+  end
+  result do
+    encoding "json"
+    field "clouds", jmes_path(response, "clouds")
+  end
+end
 datasource "ds_filtered_instances" do
-  run_script $js_filter_instances, $ds_instances, $ds_instances_tags, $param_instance_tag_list
+  run_script $js_filter_instances, $ds_instances, $ds_instances_16, $ds_instance_type
 end
 
 script "js_filter_instances", type: "javascript" do
-  parameters "instances", "instances_tags", "instance_tag_list"
+  parameters "instances", "instances16", "instance_types"
   result "filtered_instances"
   code <<-EOS
 // This is the list of filtered instances.
 var filtered_instances = [];
-
-// This is the map of instance href to its tags.
-var tags = {};
-for (var i = 0; i < instances_tags.length; i++) {
-  vt = instances_tags[i]
-  for (var j = 0; j < vt['links'].length; j++) {
-    link = vt['links'][j]
-    if (link['rel'] == 'resource') {
-      tags[link['href']] = vt['tags']
+for ( i = 0; i < instances.length; i++ ) {
+  for ( n = 0; n < instances16.length; n++) {
```

Comment:
Do we really need this nested loop here? Looks like this can be optimized. You can easily build a map/hash of the instances16 array and refer to it.

Reply:
There's even example code doing it in this repo. 😜